### PR TITLE
poetry update to get rid of security vulnerabilities, update pytest

### PR DESCRIPTION
**Pull request by olacognite**
Security updates
mypy with type hints
small changes to tests, e.g. could not pickle a local class, so had to remove the local class in test_many
Dropping support of python 3.7

### CODE REVIEW

File: cognite/processpool/__init__.py [hunk 1]
```diff
@@ -1 +1,3 @@
 from cognite.processpool.processpool import ProcessPool, ProcessPoolShutDownException, WorkerDiedException  # noqa
+
+__all__ = ["ProcessPool", "ProcessPoolShutDownException", "WorkerDiedException"]
```
Comment (cyriaquebrousse) [inline cognite/processpool/__init__.py:3]: Nice!

File: .github/workflows/test_and_build.yml [hunk 2]
```diff
@@ -11,7 +11,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [3.7, 3.8]
+        python-version: ["3.9", "3.10", "3.11"]
```
Comment (cyriaquebrousse) [inline .github/workflows/test_and_build.yml:14]: The lowest version of python you allow for the package is still 3.7, so I think we should still test as low as that

File: cognite/processpool/processpool.py [hunk 3]
```diff
@@ -129,14 +135,14 @@ def __init__(self, worker_class: type, pool_size: int = 1):
 
         self._pool = [self._create_new_worker() for _ in range(pool_size)]
 
-        self._job_queue = queue.Queue()  # type: queue.Queue # no need for a MP queue here
+        self._job_queue: queue.Queue[Optional[Job]] = queue.Queue()  # no need for a MP queue here
```
Comment (cyriaquebrousse) [inline cognite/processpool/processpool.py:138]: Why does it work here without quotation marks, but above it requires them?

File: cognite/processpool/processpool.py [hunk 4]
```diff
@@ -202,14 +208,14 @@ def _job_manager_thread(self):
                 return
             self._pool[idle_procs[0]].send(job)
 
-    def submit_job(self, *args, **kwargs) -> Future:
+    def submit_job(self, *args: Any, **kwargs: Any) -> Future[Result]:
         """Submits job asynchronously, which will eventually call the `run` method in worker_class with the arguments given, all of which should be picklable."""
         if self.terminated or self.shutting_down:
             raise ProcessPoolShutDownException("Worker pool shutting down or terminated, can not submit new jobs")
-        future = Future()  # type: Future
+        future: Future[Result] = Future()
```
Comment (cyriaquebrousse) [inline cognite/processpool/processpool.py:215]: Why does mypy need this explicitly? It should be able to handle inference here. Or is it for the generic?

File: tests/test_crashes.py [hunk 5]
```diff
@@ -18,28 +22,36 @@ def __init__(self, reason: str, response: str):
...
-    def run(self, msg, num, *args):
+    def run(self, msg: str, num: float, *args: Tuple[Any, ...]) -> float:
         time.sleep(0.1)
         if num == 1337:
-            a = []
-            while True:
-                a.extend([0] * 1000000)
+            pid = process.current_process().pid
+            assert pid is not None
+            os.kill(pid, signal.SIGKILL)
```
Comment (cyriaquebrousse) [inline tests/test_crashes.py:43]: Good change, probably more reliable!

File: cognite/processpool/processpool.py (hunk 4 above)
Comment (olacognite) [reply cognite/processpool/processpool.py:215]: For the generic, and I just preferred it in front, would probably work as a comment too

Comment (cyriaquebrousse) [reply cognite/processpool/processpool.py:215]: Between the name and the assignment is the way to go 👍🏻

File: cognite/processpool/processpool.py (hunk 3 above)
Comment (olacognite) [reply cognite/processpool/processpool.py:138]: Not sure, but this is only queue.Queue, not multiprocessing

File: .github/workflows/test_and_build.yml (hunk 2 above)
Comment (olacognite) [reply .github/workflows/test_and_build.yml:14]: Flake8 requires python >3.8.1. I think I might as well drop 3.7 at least.

File: cognite/processpool/processpool.py (hunk 3 above)
Comment (olacognite) [reply cognite/processpool/processpool.py]: Had to add the quotation marks anyway for it to work with python 3.8

**Review by cyriaquebrousse — APPROVED** [review]
Looks good to me!

**Pull request merged**